### ISSUE TITLE
Invalid key: cmd - this may mean that the key requested does not exist in your keymap

### ISSUE THREAD

**Issue by aristidesfl** [Hammerspoon/hammerspoon#1235]
Trying to toggle modifiers like command, is there a way to do this?

```
local cmdOn = hs.eventtap.event.newKeyEvent({}, 'cmd', true)
local cmdOff = hs.eventtap.event.newKeyEvent({}, 'cmd', false)
hs.hotkey.bind({}, 'f', cmdOn:post, cmdOff:post, nil)
```

**Comment by aristidesfl** [comment]
The reason why this is necessary is because something like the following is not enough to emulate ⌘⇥ application switching:

```
hs.hotkey.bind({}, 'f', keyCode('tab', {'cmd'}), nil, nil)
```

**Comment by aristidesfl** [comment]
[`CGEventCreateKeyboardEvent`](https://developer.apple.com/reference/coregraphics/1456564-cgeventcreatekeyboardevent) supports pressing modifier keys. Is there a reason why Hammerspoon doesn't?

**Comment by deryni** [comment]
You can't grab `cmdOn:post` like that. It isn't a curried function, but you could wrap it in a function easily enough.

```lua
function cmdDown()
    return cmdOn:post()
end
```

But I would imagine that you would need to send `cmd-tab` and then a synthetic `cmd`-release instead of three events ... but maybe not.

Though actually it looks like `hs.eventtap.keystrokes` is what you'd want to use here. (Though I'm not sure I see how you could get arbitrary `⌘⇥ ⇥` sequences that way.)

**Comment by aristidesfl** [comment]
The code you shared also doesn't work cause you are calling `post()` when returning. Maybe you meant something like this?
```
local function keyDown(key)
  local keyEvent = hs.eventtap.event.newKeyEvent({}, key, true)
  return function ()
    keyEvent:post()
  end
end

local function keyUp(key)
  local keyEvent = hs.eventtap.event.newKeyEvent({}, key, false)
  return function ()
    keyEvent:post()
  end
end


hs.hotkey.bind({}, 'f', keyDown('cmd'), keyUp('cmd'), nil)

```

The problem remains that `cmd` is not in the list of allowed keys. Tried with keycodes like `55` and others, but also no effect. Do you know what is the actual keycode of `cmd` key?

I actually need to manipulate `cmd` independently, since I would need for example the following sequence: `cmd:on, tab:keystroke, tab:keystroke, cmd:off` in toder to go to the second app in the list.

**Comment by deryni** [comment]
Yes, with generator functions you'd want something like that. I was writing literal functions to use as the event press and release functions (using the `cmdOn` and `cmdOff` events you pre-generated instead of generating them on the fly the way you are).

And yes, that was the thrust of my final comment. Can you create an event with the `cmd` modifier and no keys listed? Does that do what you want?

**Comment by asmagill** [comment]
at present, no, hs.eventap is not able to create a flag change event without it being a part of a key down or key up event. a planned update will support this, but i don't have a timeframe yet.

**Comment by aristidesfl** [comment]
@asmagill do you have any branch with some code? If you point me in the right direction maybe I can try to add support for it since I'm quite keen in having this feature.

**Comment by asmagill** [comment]
@aristidesfl, the code I'm working on is a complete rewrite of `hs.eventtap` in an attempt to address some slowdowns when using a lot of eventtaps and to hopefully integrate it more closely with `hs.hotkey` to address some of its limitations, so the simple answer is that it isn't a fully functional module yet ready for even alpha testing... however, I think I can merge the relevant pieces for creating flag modification events into the existing `hs.eventtap` fairly easily... I'll try and have something you can test tomorrow or Friday at the latest.

**Comment by aristidesfl** [comment]
@asmagill looking forward for that

**Comment by asmagill** [comment]
Ok, an interim `eventtap` update can be found at https://github.com/asmagill/hammerspoon_core/tree/master/eventtap.  This update modifies `hs.eventtap.event.newKeyEvent` so that it more accurately follows `CGEventCreateKeyboardEvent` and allows for modifier only key events.

Backwards compatibility for the shortcut Hammerspoon currently uses is also maintained, but you should review the README.md file which describes the "correct" way to build up key events with modifiers and migrate any code which is problematic or inconsistent to this newer approach.

@cmsj, it may be a month before I can dedicate the kind of time a proper refactoring of eventtap deserves, so if you feel this change should be introduced into core before then, let me know and I can make a pull request fairly easily.

**Comment by aristidesfl** [comment]
Thank you a lot, great documentation. I've extracted the files into `~/.hammerspoon/` and restarted Hammerspoon, reloaded config, but still receiving the same error. Am I missing something?

```sh
/Users/user/.hammerspoon/
├── hs
│   └── eventtap
│       ├── event.so
│       ├── event.so.dSYM
│       │   └── Contents
│       │       ├── Info.plist
│       │       └── Resources
│       │           └── DWARF
│       │               └── event.so
│       ├── init.lua
│       ├── internal.so
│       └── internal.so.dSYM
│           └── Contents
│               ├── Info.plist
│               └── Resources
│                   └── DWARF
│                       └── internal.so
├── init.lua
└── mobdebug.lua
```

```lua
local function keyDown(key)
  local keyEvent = hs.eventtap.event.newKeyEvent(key, true)
  return function ()
    keyEvent:post()
  end
end

local function keyUp(key)
  local keyEvent = hs.eventtap.event.newKeyEvent(key, false)
  return function ()
    keyEvent:post()
  end
end

hs.hotkey.bind({}, 'f', keyDown('cmd'), keyUp('cmd'), nil)
```

Hammerspoon 0.9.52
El Capitan 10.11.6

**Comment by aristidesfl** [comment]
Seems I didn't read your documentation thoroughly. Using `hs.eventtap.event.modifierKeys.cmd` make the error go away. Tried your example and the modifiers work fine. 

However when I put them inside functions and trigger them using `hs.hotkey.bind` they don't seem to have any effect.

Example:
```lua
local function testOn()
  print('testOn')
  hs.eventtap.event.newKeyEvent(hs.eventtap.event.modifierKeys.shift, true):post()
  hs.eventtap.event.newKeyEvent(hs.eventtap.event.modifierKeys.alt, true):post()
end

local function testOff()
  print('testOff')
  hs.eventtap.event.newKeyEvent(hs.eventtap.event.modifierKeys.alt, false):post()
  hs.eventtap.event.newKeyEvent(hs.eventtap.event.modifierKeys.shift, false):post()
end

hs.hotkey.bind({}, 'f', testOn, testOff, nil)
```

**Comment by aristidesfl** [comment]
@asmagill it seems the modifiers set in enabled in hammerspoon only affect events also sent from hammerspoon. I don't think that is a big problem.

However there seem to be a bug introduced in normal keystroke modifiers. Example:
```
hs.eventtap.keyStroke({}, 'a', 200) -- a (as expected)
hs.eventtap.keyStroke({'shift', 'alt'}, 'a', 200) -- Å (as expected)
hs.eventtap.keyStroke({}, 'a', 200) -- Å (modifiers don't get reset)
hs.eventtap.keyStroke(nil, 'a', 200) -- Å (modifiers don't get reset)
hs.eventtap.keyStroke({'alt'}, 'a', 200) -- å (as expected)
```

**Comment by asmagill** [comment]
hadn't thought to check the helper functions that use newKeyEvent, I'll take a closer look tonight.

**Comment by aristidesfl** [comment]
@asmagill found another strange behaviour which occurs when your module is loaded.

`hs.eventtap.keyStroke({}, 'delete', 500)` triggers the `Delete Left Right` macro in *Sublime Text 3*

![screen shot 2017-02-26 at 00 55 16](https://cloud.githubusercontent.com/assets/208149/23335867/77c94284-fbbe-11e6-8db1-2d3a43b079a8.png)

This macro is only supposed to be triggered when a set of conditions is true:

```json
{ "keys": ["backspace"], "command": "run_macro_file", "args": {"file": "res://Packages/Default/Delete Left Right.sublime-macro"}, "context":
    [
        { "key": "setting.auto_match_enabled", "operator": "equal", "operand": true },
        { "key": "selection_empty", "operator": "equal", "operand": true, "match_all": true },
        { "key": "preceding_text", "operator": "regex_contains", "operand": "\"$", "match_all": true },
        { "key": "following_text", "operator": "regex_contains", "operand": "^\"", "match_all": true },
        { "key": "selector", "operator": "not_equal", "operand": "punctuation.definition.string.begin", "match_all": true },
        { "key": "eol_selector", "operator": "not_equal", "operand": "string.quoted.double - punctuation.definition.string.end", "match_all": true },
    ]
},
```

Somehow, the events manage to bypass Sublime's shortcut mechanism and trigger the macro directly when `delete` is sent. Strangely this only starts to happen after one of `left, right, up, down` keystrokes have been sent. Before that, the `delete` event works as expected.

Example:
```lua
hs.eventtap.keyStroke({}, 'delete', 500)  -- normal delete
hs.eventtap.keyStroke({}, 'left', 500) -- normal left

-- from now on, delete always triggers the macro
hs.eventtap.keyStroke({}, 'delete', 500) -- Delete Left Right
hs.eventtap.keyStroke({}, 'delete', 500) -- Delete Left Right
```

**Comment by aristidesfl** [comment]
@asmagill finally got around implementing Tabmode without using hs.hotkey and using your modifiers. Unfortunately it doesn't fully work as expected and the initial problem persists. It is almost like the modifiers aren't actually pressed, but only sent along when other keycokes are sent, even though your implementation makes it look otherwise: `hs.eventtap.event.newKeyEvent(hs.eventtap.event.modifierKeys.alt, true):post()`. Is this the case?

When I send the second tab event to go to subsequent applications in the list, the command key either gets released, or is no longer there, so the application switcher disappears. Or perhaps it is something else?

Here is a very rough video demonstrating the problem.
https://youtu.be/8405n9CowvA

**Comment by asmagill** [comment]
Sorry for the delay... I have not had a chance yet to adjust the helper functions, so presently I would only trust `hs.eventtap.event.newKeyEvent` with the new keys.

My understanding, and this is just from my observation (if anyone knows a source that can confirm/deny/expand this, please share), is that the `flagsChanged` events are sent as you would expect, but they are only merged with `keyUp` and `keyDown` events generated from the same source as the `flagsChanged` event... what this means is that if we generate a `flagsChanged` event for the `shift` key, then only `keyDown` and `keyUp` events generated by Hammerspoon will inherit this `shift` key (until we send a `flagsChanged` event clearing it).  This means that we can't (unfortunately... I tried) cause key up/down events from typing on the keyboard to inherit this shift... unless we create an eventtap which intercepts these events from the keyboard and then generates its own replacement event, which then *would* inherit the `shift`.

I have not had a chance to review your code specifically or the video yet but I will before I submit a pull that includes updating the helper functions later this week and see if i can either shed some light, make some suggestions, or modify the code if necessary/possible.

**Comment by aristidesfl** [comment]
Thank you @asmagill 

To make it cleared what is the situation at the moment:

1. I'm only using `hs.eventtap.event.newKeyEvent` to emit new events. 
2. The first problem is a weird bypass/triggering of shortcut in Sublime Text 3
3. The other is that using your mod to enable modifiers, they seem to get unpressed (according to the keyboard viewer) as soon as a subsequent event is sent. Maybe this or something else is causing the application switcher to disappear.

**Comment by asmagill** [comment]
I have some time now to take a closer look at this...

First, it appears that you are correct... when the shortcut version is used (i.e. with a table containing one or more modifiers is included) these do not get "unset" after the event... I have an idea why but the  simple fix is going to break backwards compatibility... I need to think about it and test some more...

Any chance of seeing the full code you're trying to create?  I can get most of the gist from the video you provide, but having a copy I can hammer on myself might help.

I'll have to check out Sublime Text or see if I can find something I already have that uses the delete key as a shortcut.  If you were trying to trigger `Delete Left Right` from Sublime Text normally from the keyboard, what specific combination of keys/modifiers trigger it?  I'm assuming not a simple tap of the delete key or this behavior wouldn't have surprised you.

**Comment by asmagill** [comment]
An updated version of eventtap is at https://github.com/asmagill/hammerspoon_core/tree/master/eventtap.  The only real difference is that `newKeyEvent` and `keyStroke` now behave differently if the modifiers table is `{}` (i.e. an empty table) or `nil`... for the empty table, it will "reset" any currently posted modifiers that are in the "down" state while `nil` causes the event to "inherit" any currently engaged modifiers.  The modifiers parameter is still optional (defaults to nil) for `newKeyEvent`, but still required for `keyStroke` (though it can be nil, as described here).

**Comment by aristidesfl** [comment]
# Tabmode
I've tried your new v0.2 but both the problem with sublime and tabmode persist. Here is the code if you can have a look:  https://gist.github.com/aristidesfl/e695bff729b96b1223faeda5287f81b6

In addition to that, I've started to try implementing the functionality in Swift in this [repository](https://developer.apple.com/reference/coregraphics/cgevent/1456564-init#discussion). To my surprise it suffers from the same problem as hammerspoon. It doesn't behave has the [documentation ](https://developer.apple.com/reference/coregraphics/cgevent/1456564-init#discussion) suggests. Maybe you have clue why? There is an open [question](https://stackoverflow.com/questions/17773225/cgeventpost-hold-a-key-shift) on Stackoverflow as a well.

# Sublime + Delete keystroke

> If you were trying to trigger Delete Left Right from Sublime Text normally from the keyboard, what specific combination of keys/modifiers trigger it?

This macro comes with sublime and takes care of removing matching pairs of `()[]{}` when you delete the first. It is triggered automatically by the delete key only, but only in the right context (between empty brackets for example). Here is the conditions which must be met: 
```js
{ "key": "setting.auto_match_enabled", "operator": "equal", "operand": true },
{ "key": "selection_empty", "operator": "equal", "operand": true, "match_all": true },
{ "key": "preceding_text", "operator": "regex_contains", "operand": "\"$", "match_all": true },
{ "key": "following_text", "operator": "regex_contains", "operand": "^\"", "match_all": true },
{ "key": "selector", "operator": "not_equal", "operand": "punctuation.definition.string.begin", "match_all": true },
{ "key": "eol_selector", "operator": "not_equal", "operand": "string.quoted.double - punctuation.definition.string.end", "match_all": true }
```

**Comment by asmagill** [comment]
Sorry for the delay -- I haven't forgotten about this, but I haven't had much time to plug at it recently... any changes or updates?  I will review the code at the link you link to above, but just wanted to see if you have any new thoughts or changes before I get too deep into it.

**Comment by aristidesfl** [comment]
@asmagill I ended up implementing the whole thing in Swift. I will share the source in a couple of days. Basically the issue, is that the system will release a modifier a soon as a subsequent keyUp or keyDown event doesn't contain the modifier flags. This is not what is described in the Apple documentation, which might be outdated. So for `⌘⇥⇥⇥⌘` as an example, you would have to to send keyDown for ⌘, and then still turn on the ⌘ flags on each ⇥ sent.

**Comment by aristidesfl** [comment]
Regarding the Sublime + Hammerspoon problem, I don't know what could be causing it. I can imagine that it might be related to the layer where the events are inserted. Maybe they are inserted at a layer that bypasses Sublimes shortcut mechanisms, but still activates macOS menu item keyboard shortcut system.

**Comment by asmagill** [comment]
Regarding tabmode:

tl;dr - I'd like to see your Swift code when you're ready, especially if you've got it working... I'm really curious how.

As to my observations of your lua version, see below:

First observation, your eventwatchers at lines 67 and 77 are not saved in any global variable (or local variable tied up in an up-value)... this means that the first time lua does garbage collection, they go bye-bye.

Second observation, after assigning the eventtaps to global vars so they don't get collected and using the following to observe what events are actually occurring, it seems that our (Hammerspoon's) keypresses also have a `eventSourceUnixProcessID` of 0 -- did not expect that myself either...

~~~lua
flatinspect = function(what)
    return (hs.inspect(what, {
        process = function(i,p)
            if p[#p] == "flags" then
                 -- supresses the nonCoalesced modifier and what seems to be a synthesized event flag, though
                 -- I haven't found supporting documentation to confirm this.  These two modifiers are not
                 -- something we can control/set and they just confuse things when comparing against the values
                 -- defined in IOLLEvent.h
                return string.format("0x%08x", i & 0xdffffeff)
            elseif p[#p] == "type" then
                -- displays the event type rather than its number
                return hs.eventtap.event.types[i] or i
            elseif p[#p] == "keycode" then
                -- displays the actual key instead of the keycode number
                return hs.eventtap.event.modifierKeys[i] or hs.keycodes.map[i] or i
            else
                return i
            end
        end
    }):gsub("%s+", " ")) -- the extra () wrapping the return value suppresses string.gsub's second return value
end

a = hs.eventtap.new({
    hs.eventtap.event.types.flagsChanged,
    hs.eventtap.event.types.keyUp,
    hs.eventtap.event.types.keyDown
}, function(e)
    if not (e:getType() == hs.eventtap.event.types.keyDown and e:getProperty(hs.eventtap.event.properties.keyboardEventAutorepeat) ~= 0) then
        print(timestamp(), string.format("%s, %d, 0x%x", flatinspect(e:getRawEventData().CGEventData), e:getProperty(hs.eventtap.event.properties.eventSourceUnixProcessID), e:getProperty(hs.eventtap.event.properties.eventSourceUserData)))
    end
end):start()
~~~

I further modified your code so that `keystroke` reads:

~~~lua
local function keyStroke(key, modifiers)
  return function()
    anyKeyPressed = true
    toggleModifiers(modifiers, true)
    hs.eventtap.event.newKeyEvent(key, true):setProperty(hs.eventtap.event.properties.eventSourceUserData, 0xc0ffee):post()
    hs.eventtap.event.newKeyEvent(key, false):setProperty(hs.eventtap.event.properties.eventSourceUserData, 0xc0ffee):post()
    toggleModifiers(modifiers, false)
  end
end
~~~

and changed the check in both eventtaps to

~~~lua
if event:getKeyCode() == hs.keycodes.map[activator] and event:getProperty(hs.eventtap.event.properties.eventSourceUserData) ~= 0xc0ffee then
~~~

and it kinda-sorta works... for one switch only... at a guess, it looks like the code which displays the application switcher is checking a different event session than we are using to see if it should treat the next `tab` as a continuation or a brand new tab (if you look closely at the messages printed by my watcher, you'll see that the events which your code generates *do* maintain the `cmd` key being pressed, but that the ones that correspond to actual keyboard presses do not).

FWIW, Hammerspoon uses the source session generated by `CGEventSourceCreate(kCGEventSourceStatePrivate)`... if I hard code Hammersopon to use `CGEventSourceCreate(kCGEventSourceStateCombinedSessionState)`, then the `cmd` gets released as soon as you tap the `l` key (and who knows what other behaviors people have come to expect might break... I may allow the user to choose between the two in the update that I keep trying to get back to, but so far, I've found much that breaks when using the combined state and nothing that's gained, at least so far in my tests)

**Comment by asmagill** [comment]
Regarding Sublime... at a guess, and I've never used Sublime so that's what this is -- a guess... You're probably pretty close to right... 

tl;dr -- Short version, unless we can find a way to post the events to a specific point in the application's responder chain (and I'm pretty sure we can't, at least not without private methods, and even then I'm not sure), I'm not sure we can do anything about it.

An NSTextView can intercept keypresses before the application's event loop, and most windows which allow major text editing use some variant or subclass of NSTextView (or similar).  What we're sending gets sent directly to the applications event loop, which (I think) bypasses the responder chain, so the window or view never has an opportunity to intercept the key event and perform its own logic.

Again, this is conjecture based upon my observations with other applications and work I've been toying with off and on to create a module which would allow us to create editable text windows within Hammerspoon, so... if anyone knows more or otherwise, I'm all ears!

OTOH if you know of an application that can, especially if it's open source, let me know and I'll try and figure out what they're doing that we're not.

**Comment by aristidesfl** [comment]
@asmagill the code is a bit rough, since it was a PoC, I didn't have time to refactor and no previous experience with Swift and macOS framekworks. But it is working. Also there is a lot of code which is specific to the modes I implemented, so the code is not a generic implementation of modals, etc.

### Regarding Tab Mode
1. https://github.com/aristidesfl/octopus/blob/master/octopus/AppDelegate.swift#L79-L86
2. https://github.com/aristidesfl/octopus/blob/master/octopus/AppDelegate.swift#L96-L97

### Regarding Sublime and where to insert the events: https://github.com/aristidesfl/octopus/blob/master/octopus/Keyboard.swift#L16
I think I tried other possible values there and that is the one which worked the best.

**Comment by asmagill** [comment]
Leaving this open because I want to take a closer look at @aristidesfl's links above and see if it suggests further additions.
